src/librustc/traits/error_reporting.rs

```diff
@@ -837,6 +832,47 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
         err
     }
+
+    fn report_closure_arg_mismatch(&self,
+                           span: Span,
+                           found_span: Option<Span>,
+                           expected_ref: ty::PolyTraitRef<'tcx>,
+                           found: ty::PolyTraitRef<'tcx>)
+        -> DiagnosticBuilder<'tcx>
+    {
+        fn build_args_type_string(trait_ref: &ty::TraitRef) -> String {
+            match trait_ref.substs.type_at(1).sty {
+                ty::TyTuple(args, _) => {
+                    let len = args.len();
+                    if len == 0 {
+                        return String::from("no arguments");
+                    }
+                    let mut type_list_string = String::new();
+                    if len >= 2 {
+                        for ty in args[..len-2].iter() {
+                            type_list_string.push_str(&format!("`{}`, ", ty));
+                        }
+                        type_list_string.push_str(&format!("`{}` and ", args[len-2]));
+                    }
+                    type_list_string.push_str(&format!("`{}`", args[len-1]));
+                    format!("argument{} of type {}",
+                            if len == 1 { "" } else { "s" },
+                            type_list_string)
+                },
+                _ => panic!(),
```

Comment:
Is it fine to use standard panic macro like `panic!` or `unreachable!` to signal compiler bugs? (`panic!` here is a mistake, I should have used `unreachable!` here)

Reply:
An `unreachable!()` is ok-ish for local things, but it is recommended to use `bug!` or `span_bug!(span, "error")` - `span_bug` also displays the span of the error, which makes for easier debugging.